include/engine/base64.hpp

```diff
@@ -16,12 +17,26 @@
 
 // RFC 4648 "The Base16, Base32, and Base64 Data Encodings"
 // See: https://tools.ietf.org/html/rfc4648
-// Implementation adapted from: http://stackoverflow.com/a/28471421
 
+namespace detail
+{
 // The C++ standard guarantees none of this by default, but we need it in the following.
 static_assert(CHAR_BIT == 8u, "we assume a byte holds 8 bits");
 static_assert(sizeof(char) == 1u, "we assume a char is one byte large");
 
+using Base64FromBinary = boost::archive::iterators::base64_from_binary<
+    boost::archive::iterators::transform_width<const char *, // sequence of chars
+                                               6,            // get view of 6 bit
+                                               8             // from sequence of 8 bit
+                                               >>;
+
+using BinaryFromBase64 = boost::archive::iterators::transform_width<
+    boost::archive::iterators::binary_from_base64<std::string::const_iterator>,
+    8, // get a view of 8 bit
+    6  // from a sequence of 6 bit
+    >;
+} // ns detail
```

Comment:
these two transform_widths require the stream length to be a multiple of 24bits / 3 bytes, that's why we pad it _before_ using them for iteration!